pkg/recommendation/device.go

```diff
@@ -0,0 +1,250 @@
+package recommendation
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"mayadata.io/cstorpoolauto/types"
+	bdutil "mayadata.io/cstorpoolauto/util/blockdevice"
+)
+
+// cStorPoolClusterRecommendationRequest is used to request the device
+// recommendation for a given raid type.
+type cStorPoolClusterRecommendationRequest struct {
+	Request types.CStorPoolClusterRecommendationRequest
+	Data    Data
+}
+
+// Data is the input data requested for the specifix recommendation.
+type Data struct {
+	BlockDeviceList *unstructured.UnstructuredList
+}
+
+// NewRequestForDevice returns a device request object after validation.
+func NewRequestForDevice(request *types.CStorPoolClusterRecommendationRequest, data *Data) (
+	*cStorPoolClusterRecommendationRequest, error) {
+
+	if request.Spec.PoolCapacity.IsZero() {
+		return nil, errors.New(
+			"Unable to create device recommendation request: Got zero pool capacity")
+	}
+	if data.BlockDeviceList == nil {
+		return nil, errors.New(
+			"Unable to create device recommendation request: Got nil block device list")
+	}
+
+	err := request.Spec.DataConfig.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to create capacity recommendation request")
+	}
+
+	cspcrr := cStorPoolClusterRecommendationRequest{
+		Request: types.CStorPoolClusterRecommendationRequest{
```

Comment:
why not
```go
Request: *request
```

Reply:
Why is pointer needed?

Reply:
just to check the client is providing or not. Basically nil check is better than empty check

Reply:
So if you check I am accepting pointer only: 
```
if request.Spec.PoolCapacity.IsZero() {
		return nil, errors.New(
			"Unable to create device recommendation request: Got zero pool capacity")
	}
```
here request argument is pointer only.